---
title: "Eatomics Quality Control and Differential Abundance Report"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: 
  html_document:
    theme: flatly
params:
  linesFromMaxQuant: NA
  stats_proteinGroups: NA
  pca: NA
  TSdetect: NA
  norm: NA
  number: NA 
  coverage: NA 
  missval: NA 
  detect: NA 
  impute : NA 
  StSheatmap : NA 
  StSheatmapDistMetric: NA
  cumsum : NA 
  distributionPlot : NA
  volcano_plot: NA
  ExpSetup: NA
  UpRegul: NA
  DoRegul: NA
  boxPlotUp: NA
  boxPlotDown: NA
 
---

## 1. Data description, preprocessing and quality control

The MaxQuant algrorithm identified `r params$linesFromMaxQuant` proteins across all samples. 
`r params$stats_proteinGroups$NumPotCon` potential contaminants, `r params$stats_proteinGroups$IdentifiedBySite` proteins only identified by site, and `r params$stats_proteinGroups$Reverse` proteins only identified in the reverse database were removed from the data set. The final dataset has `r params$stats_proteinGroups$Cleaned` rows. `r params$stats_proteinGroups$IntensityMetric` was used as intensity metric for quality control plots and differential expression analysis.
Missing values are imputed using the `r params$stats_proteinGroups$Imputation` method. Imputation is based on knn (k-nearest-neighbour) from the impute package [1], MinDet or QRLIC from the imputeLCMD package with default parameters [2] or a custom implementation of Perseus' sampling from a down-shifted Gaussian distribution (implemented by Matthias Ziehm) with default parameters of width = 0.3 and shift = 1.8.
The plots on protein counts per sample and the sample coverage describe the proteomic depth on a per sample level. 
The detection of highly abundant proteins can be examined in the cumulative intensities plot, while low abundance proteins are identified in the low abundance transcription factors plot.

### 1.1 Protein counts per sample

Protein numbers describes the count of distinct proteins or isoforms per sample. The Eatomics::plot_proteinCoverage() function is used to generate the plot.

```{r number, echo=F, fig.width=10}
if(is.null(params$number)){
  message("The protein numbers plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$number
}
```

### 1.2 Cumulative protein intensities

Protein intensities are cumulated across all samples and plotted according to their relative abundance. Colouring marks the respective quantile of the proteins. Highly abundant proteins, i.e., proteins ranked in the first quartile are coloured in red and labels are specified. The top 20 ranked proteins and their cumulated intensity are given in the table to the right. The Eatomics::plot_CumSumIntensities() function is used to generate the plot.

```{r cumsum, echo = F}
library(gridExtra)
if(is.null(params$cumsum)){
  message("The missing values imputation plot was not computed during the analysis and thus is not part of this report.\n")
} else {
 #gridExtra::grid.draw(params$cumsum)
  params$cumsum
}
```

### 1.3 Sample to sample heatmap

The sample-to-sample heatmap describes the biological and/or technical variability of the samples. It can be produced using different (dis-) similarity metrics (pearson correlation or euclidean distance). For this plot ´r params$StSheatmapDistMetric´ was used. Formed clusters should resemble the sample groups under investigation. The Eatomics::plot_StS_heatmap() function is used to generate the plot. plot_StS_heatmap

```{r stsheatmap, echo=F, fig.height=10}

if(is.null(params$StSheatmap)){
  message("The sample to sample heatmap was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$StSheatmap
}
```


### 1.4 Principal component analysis

A common method of dimensionality reduction is principal component analysis (PCA). Inherently,
PCA calculates axes of most variation (principal components) within the expression data. A common assumption is that a plot along the axes of most variation will seggregate all samples/patients into groups under investigation. The stats::prcomp() function is used to calculate PCA data. 

```{r pca, echo = F}
if(is.null(params$pca)){
  message("The PCA was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$pca
}


```

### 1.4 Data distribution

Boxplots showing the sample-wise data distribution of log2 transformed intensity values. Colours are user selected. 
By default, normalization via limma::normalizeVSN() is done for iBAQ values whereas LFQ values are assumed to be normalized already. 


```{r distrib, echo = F}

if(is.null(params$distributionPlot)){
  message("The normalization plot was not computed during the analysis and thus is not part of this report.\n")
} else {
 params$distributionPlot
}

```

### 1.7 Missing Values - Quantitative 

The missing values plot visualizes the density and cumulative protein intensities with and without missing values. It can guide the selection of the missing value imputation strategy. The plot is generated by the plot_impute() function from the DEP package which was adjusted to work without experimental design information. 

```{r detect,  echo = F}
#library(gridExtra)

#grid.arrange(params$detect,top="Main Title")
if(is.null(params$detect)){
  message("The missing values plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$detect
}
```


### 1.9 Imputation
The missing value imputation plot shows the density of missing and valid values over range of the spectrun of intensities before and after imputation with the selected method. The plot is generated by the plot_detect() function from the DEP package which was adjusted to work without experimental design information. 

```{r impute, echo = F}

if(is.null(params$impute)){
  message("The missing values imputation plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$impute
}

```


## 2. Limma Differential Abundance Results

### 2.1 Volcano plot
Differential expression volcano plot, x axis plotting log2 FoldChange (logFC) and y axis plotting -log10 of P.value from the limma [`toptable` function](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/toptable). P-values are multiple testing corrected by using the [Benjamini-Hochberg procedure](https://www.rdocumentation.org/packages/stats/versions/3.5.2/topics/p.adjust). 

```{r Limma Output, echo=F}

if(is.null(params$volcano_plot)){
  message("The volcano plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$volcano_plot
}
```

### 2.2 Boxplots of selected protein abundance

User selected proteins showing higher abundance. 

```{r Limma BoxplotUp, echo=F}

if(is.null(params$boxPlotUp)){
  message("The boxplot for proteins significantly higher in abundance was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$boxPlotUp
}
```

User selected proteins showing lower abundance.  

```{r Limma BoxplotDown, echo=F}

if(is.null(params$boxPlotDown)){
  message("The boxplot for proteins significantly lower in abundance was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$boxPlotDown
}
```

### 2.3 Table of significant proteins
To see the full list of Up and Down regulated gene list, jump to [table of significant differences (up)](#appendix-Up) and [table of significant differences (down)](#appendix-Down), respectively.

### 2.4 Detailed Discription

```{r Experimental Setup, echo=F}
params$ExpSetup
```

## References

[1]  Trevor Hastie, Robert Tibshirani, Balasubramanian Narasimhan and Gilbert Chu (2018). impute: impute: Imputation for
  microarray data. R package version 1.56.0.
  
[2] Cosmin Lazar (2015). imputeLCMD: A collection of methods for left-censored missing data imputation. R package version 2.0.
  https://CRAN.R-project.org/package=imputeLCMD

## Appendix 

### List of proteins with higher abundance {#appendix-Up}

```{r Upregul, echo=F }
library("kableExtra")
kable(params$UpRegul, caption = " ") %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed")) 
```

### List of proteins with less abundance {#appendix-Down}

```{r DoRegul, echo=F }
library("kableExtra")
kable(params$DoRegul) %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed")) 
```

