---
title: "Eatomics Proteome Exploration Report"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: 
  html_document:
    theme: flatly
params:
  linesFromMaxQuant: NA
  stats_proteinGroups: NA
  pca: NA
  TSdetect: NA
  norm: NA
  number: NA 
  coverage: NA 
  missval: NA 
  detect: NA 
  impute : NA 
  StSheatmap : NA 
  StSheatmapDistMetric: NA
  cumsum : NA 
  volcano_plot: NA
  ExpSetup: NA
  UpRegul: NA
  DoRegul: NA
  boxPlotUp: NA
  boxPlotDown: NA
  gsea_volcano_plot: NA
  gseaSetup: NA
  gseaUpRegul: NA
  gseaDoRegul: NA
  separateValuesGSEA: NA
 
---

## 1. Data description and quality control
The MaxQuant algrorithm identified `r params$linesFromMaxQuant` proteins across all samples. 
`r params$stats_proteinGroups[[2]]` potential contaminants, `r params$stats_proteinGroups` proteins only identified by site, and `params$stats_proteinGroups[[3]]` proteins only identified in the reverse database were removed from the data set. The final dataset has `r params$stats_proteinGroups` rows. 
The plots on protein counts per sample and the sample coverage describe the proteomic depth on a per sample level. 
The detection of highly abundant proteins can be examined in the cumulative intensities plot, while low abundance proteins are identified in the low abundance transcription factors plot.

### 1.1 Protein counts per sample

Protein numbers describes the count of distinct proteins or isoforms per sample. The plot_proteinCoverage() funtion is used to generate the plot.

```{r number, echo=F, fig.width=10}
if(is.null(params$number)){
  message("The protein numbers plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$number
}
```


### 1.2 Sample Coverage

Sample coverage describes how many proteins are shared across samples. I.e. there will be proteins that were detected in only one sample (top of the plot) or proteins detected in all samples (bottom of the plot). The plot is generated by the plot_coverage() function from the DEP package which was adjusted to work without experimental design information. 

```{r coverage, echo = F}

if(is.null(params$coverage)){
  message("The sample coverage plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$coverage
}
```

### 1.3 Cumulative Protein Intensities

Protein intensities are cumulated across all samples and plotted according to their relative abundance. Colouring marks the respective quantile of the proteins. Highly abundant proteins, i.e., proteins ranked in the first quartile are coloured in red and labels are specified. The top 20 ranked proteins and their cumulated intensity are given in the table to the right.

```{r cumsum, echo = F}
library(gridExtra)
if(is.null(params$cumsum)){
  message("The missing values imputation plot was not computed during the analysis and thus is not part of this report.\n")
} else {
 #gridExtra::grid.draw(params$cumsum)
  params$cumsum
}
```

### 1.4 Detection of low abundance transcription factors

Gene lists for the families GATA, JUN, SH2, SOX and TBX were downloaded from https://www.genenames.org on 23.03.2017. Please note, that in fact not all members of these families are neccessarily low-abundant. 

```{r TSdetect, echo=F, fig.height=10}
if(is.null(params$TSdetect)){
  message("The detection of low abundance transcription factors was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$TSdetect
}

```


### 1.5 Sample to sample heatmap

The sample-to-sample heatmap describes the biological and/or technical variability of the samples. It can be produced using different (dis-) similarity metrics (pearson correlation or euclidean distance). For this plot ´r params$StSheatmapDistMetric´ was used. Formed clusters should resemble the sample groups under investigation. 

```{r stsheatmap, echo=F, fig.height=10}

if(is.null(params$StSheatmap)){
  message("The sample to sample heatmap was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$StSheatmap
}
```


### 1.6 Principal component analysis

A common method of dimensionality reduction is principal component analysis (PCA). Inherently,
PCA calculates axes of most variation (principal components) within the expression data. A common assumption is that a plot along the axes of most variation will seggregate all samples/patients into groups under investigation. 

```{r pca, echo = F}
if(is.null(params$pca)){
  message("The PCA was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$pca
}


```


```{r norm, eval=FALSE, include=FALSE}

### 1.7 Normalization

#Normalization in EAtomics can be performed via the DEP package which in turn uses normalize_vsn(), a variance stabilizing transformation using the vsn-package. A comparison of data before and after normalization is shown in the plot. The plot is generated by the plot_normalization() function from the DEP package which was adjusted to work without experimental design information. By default, normalization is done for iBAQ values whereas LFQ values are assumed to be normalized already. 

if(is.null(params$norm)){
  message("The normalization plot was not computed during the analysis and thus is not part of this report.\n")
} else {
 params$norm
}

```

### 1.7 Missing Values - Quantitative 

The missing values plot visualizes the density and cumulative protein intensities with and without missing values. It can guide the selection of the missing value imputation strategy. The plot is generated by the plot_impute() function from the DEP package which was adjusted to work without experimental design information. 
```{r detect,  echo = F}
#library(gridExtra)

#grid.arrange(params$detect,top="Main Title")
if(is.null(params$detect)){
  message("The missing values plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$detect
}
```

### 1.8 Missing Values - Heatmap

The missing values heatmap shows the distribution of missing values across samples. White areas mark missing values, while black is a valid value. If samples cluster based on their pattern of missing values, caution should be taken in further analysis as this pattern may perturb, e.g., differential expression analysis. In well-founded cases, samples should be removed from the analysis. The plot is generated by the plot_missval() function from the DEP package. 

```{r missval,  echo = F}

if(is.null(params$missval)){
  message("The missing values heatmap was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$missval
}
```

### 1.9 Imputation
The missing value imputation plot shows the density of missing and valid values over range of the spectrun of intensities before and after imputation with the selected method. The plot is generated by the plot_detect() function from the DEP package which was adjusted to work without experimental design information. 

```{r impute, echo = F}

if(is.null(params$impute)){
  message("The missing values imputation plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$impute
}

```


## 2. Limma Differential Expression Results

### 2.1 Volcano plot
Differential expression volcano plot, x axis plotting log2 FoldChange (logFC) and y axis plotting -log10 of P.value from the limma [`toptable` function](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/toptable) .

```{r Limma Output, echo=F}

if(is.null(params$volcano_plot)){
  message("The volcano plot was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$volcano_plot
}
```

### 2.2 Boxplots of Proteins with lowest p.adj values

User selected upregulated protein expression. 

```{r Limma BoxplotUp, echo=F}

if(is.null(params$boxPlotUp)){
  message("The boxplot for most significantly upregulated proteins was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$boxPlotUp
}
```

User selected downregulated protein expression. 

```{r Limma BoxplotDown, echo=F}

if(is.null(params$boxPlotDown)){
  message("The boxplot for most significantly upregulated proteins was not computed during the analysis and thus is not part of this report.\n")
} else {
  params$boxPlotDown
}
```

### 2.3 Up and Down Regulated Genelist Table
To see the full list of Up and Down regulated gene list, jump to [appendix-Upregulated Genelist](#appendix-Up) and [appendix-Downregulated Genelist](#appendix-Down), respectively.

### 2.4 Detailed Discription

```{r Experimental Setup, echo=F}
params$ExpSetup
```

## 3. Differential ssGSEA Pathway Analysis (`r params$separateValuesGSEA[1]`)

### 3.1 Volcanoplot

```{r GSEA Output, echo=F}
params$gsea_volcano_plot

```

### 3.2 Up and Down Regulated Pathway  Table
To see the full list of Up and Down regulated geneset pathways, jump to [appendix-Upregulated Pathways](#appendix-gseaUp) and [appendix-Downregulated Pathways](#appendix-gseaDown), respectively.

### 3.3 Detailed Discription

```{r DiffGSEA Experimental Setup, echo=F}
params$gseaSetup
```

## Appendix 

### Upregulated Genelist {#appendix-Up}

```{r Upregul, echo=F }
library("kableExtra")
kable(params$UpRegul, caption = " ") %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed")) 
```

### Downregulated Genelist{#appendix-Down}

```{r DoRegul, echo=F }
library("kableExtra")
kable(params$DoRegul) %>% kable_styling(bootstrap_options = c("striped", "hover", "condensed")) 
```

### Upregulated Pathways {#appendix-gseaUp}

```{r pathway_Upregul, echo=F }
library("kableExtra")
#library("seplyr")
kable(params$gseaUpRegul) %>% kable_styling(bootstrap_options = c("condensed", "striped", "bordered"))
```

### Downregulated Pathways {#appendix-gseaDown}

```{r pathway_DoRegul, echo=F }
library("kableExtra")
#library("seplyr")
kable(params$gseaDoRegul) %>% kable_styling(bootstrap_options = c("condensed", "striped", "bordered"))
```
